### DISCUSSION TITLE
only run workflow if pushed tag is on latest commit

### DISCUSSION THREAD

**Discussion by sku0x20** [community/community#40625] [Actions] [answered]
### Select Topic Area

Question

### Body

i am trying to reuse deploy-jar tag. 

whenever someone pushes deploy-jar tag, release workflow is ran and this tag is deleted from repo.
since this tag is deleted on remote but not in the local repo. on next push this tag is again pushed. and the workflow is ran again.

is there a way to not run this when this happens. or to only run when the tag is pushed on latest commit.

**Comment by airtower-luna** [comment, marked as answer]
No, there isn't anything built-in, and there are some problems with the logic:

* There is no "latest commit" as such. You could write your own logic to compare the tag with a *specific branch* and run the rest of the workflow only if it matches, but:
* The general expectation in Git is that a tag points at a specific commit, *and then never changes*. See the [`git tag` documentation section "On Re-Tagging"](https://git-scm.com/docs/git-tag#_on_re_tagging).

If you want to tell actions to deploy from a certain commit by pushing (there are other options), the logical thing would be to use a specific *branch* and push the commit you want to that branch.

**Reply by sku0x20** [reply]
thank you. 

having a release branch is a good idea. still i want to avoid having a different branch. 

can you give me a bit info about the first one. where to put my logic in. how to cancel a workflow. 
also a bit about which git commands to use. 
i tried searching but could not find anything or didn't knew where/what to look for, i could not put up all the pieces at single place. 🙃

i am trying to use tag just as a trigger mechanism for release, tags created by releasing are following the git principles.

or should i use a manual trigger. this would require gh on the developer's machine. but it would solve this?

anyways, thank you... :)

ps: discussions are the best. i love them so much. makes me realize there are so many better ways... :)

**Reply by imen641** [reply]
.

**Comment by airtower-luna** [comment]
> having a release branch is a good idea. still i want to avoid having a different branch.

There's no harm in having a lot of branches, they cost almost nothing (in terms of storage and CPU time). However, from the rest of your description I don't think you need one.

> can you give me a bit info about the first one. where to put my logic in. how to cancel a workflow.
> also a bit about which git commands to use.

For logic, I'd make a separate job that checks the condition, and creates a job output that says whether the rest should run. Then make the other job(s) depend on that one (`needs`), and add a job-level `if` that checks if it should run. If you want to compare if two refs point at the same commit, you can use `git rev-parse` on both and check if the result is the same.

> i am trying to use tag just as a trigger mechanism for release, tags created by releasing are following the git principles.

This is what makes me think you don't need any special branch or tag. Just create the release tag, push it, and let the release logic run from there. Super common approach, and simple. :slightly_smiling_face: 

> or should i use a manual trigger. this would require gh on the developer's machine. but it would solve this?

Well, you could [use a `workflow_dispatch`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch). The dispatch is aimed at a ref (branch or tag), but considering that you only want to deploy if the commit is head of a certain branch you can just have the workflow run directly for the branch.

**Reply by sku0x20** [reply]
again thank you.

> This is what makes me think you don't need any special branch or tag. Just create the release tag, push it, and let the release logic run from there. Super common approach, and simple. 🙂

my release logic is really stupid. the release tag is based on date. and/but i don't want the tag/version to be developers machine date.
it should be a common source. so i am using github runners date.

> Well, you could [use a workflow_dispatch](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch). The dispatch is aimed at a ref (branch or tag), but considering that you only want to deploy if the commit is head of a certain branch you can just have the workflow run directly for the branch.

thanks, i would try to use this. i just need to change my gradle deploy task to use gh cli to trigger the run instead of tagging... :)

what i didn't thought of is, i tag the commit push the tag and then remove the tag from local repo. in this case i would just have to add another git command to remove the tag in the deploy gradle task. is this that simple. oh man. lol... :)

anyways, thank you for the guidance...

edit: only issue with deleting tag is, if someone pulls when the release action is queued. then he/she will get the deploy tag. that is quite rare but that could happen. it get's the job done but it's not reliable, triggering from gh cli seems much better option...

**Reply by airtower-luna** [reply]
> my release logic is really stupid. the release tag is based on date. and/but i don't want the tag/version to be developers machine date.
> it should be a common source. so i am using github runners date.

Just use a fixed timezone for calculating the date, and you have the same thing without the convoluted process. UTC would be the obvious choice. :wink: 

If developers' machines are more than a few min out of sync you have a massive problem anyway… :sweat_smile:

**Reply by sku0x20** [reply]
easier said than done. i haven't looked but on windows it's pain in the ass. linux can use shell substitution and what not, but windows *\*sigh\**. 
just for this i don't want to go and write a program and stuff...

this is cheap, yet reliable solution. at least for now... :)

**Reply by urica12** [reply]
Okay

**Comment by imen641** [comment]
.

**Comment by urica12** [comment]
Push
